apollo-federation/src/sources/connect/json_selection/README.md

```diff
@@ -574,25 +637,77 @@ registered in the `JSONSelection` parser in order to be recognized.
 For the time being, only a fixed set of known methods are supported, though this
 list may grow and/or become user-configurable in the future:
 
-> Full disclosure: even this list is still aspirational, but suggestive of the
-> kinds of methods that are likely to be supported in the next version of the
-> `JSONSelection` parser.
-
 ```graphql
-list->first { id name }
-list->last.name
-list->slice($args.start, $args.end)
-list->reverse
-some.value->times(2)
-some.value->plus($addend)
-some.value->minus(100)
-some.value->div($divisor)
-isDog: kind->eq("dog")
-isNotCat: kind->neq("cat")
-__typename: kind->match({ "dog": "Dog", "cat": "Cat" })
-decoded: utf8Bytes->decode("utf-8")
-utf8Bytes: string->encode("utf-8")
-encoded: bytes->encode("base64")
+# The ->echo method returns its first input argument as-is, ignoring
+# the input data. Useful for embedding literal values, as in
+# $->echo("give me this string"), or wrapping the input value.
+__typename: $->echo("Book")
+wrapped: field->echo({ fieldValue: @ })
+
+# Returns the type of the data as a string, e.g. "object", "array",
+# "string", "number", "boolean", or "null". Note that `typeof null` is
+# "object" in JavaScript but "null" for our purposes.
+typeOfValue: value->typeof
+
+# When invoked against an array, ->map evaluates its first argument
+# against each element of the array, binding the element values to `@`,
+# and returns an array of the results. When invoked against a non-array,
+# ->map evaluates its first argument against that value and returns the
+# result without wrapping it in an array.
+doubled: numbers->map(@->mul(2))
+types: values->map(@->typeof)
+
+# Returns true if the data is deeply equal to the first argument, false
+# otherwise. Equality is solely value-based (all JSON), no references.
+isObject: value->typeof->eq("object")
+
+# Takes any number of pairs [candidate, value], and returns value for
+# the first candidate that equals the input data. If none of the
+# pairs match, a runtime error is reported, but a single-element
+# [<default>] array as the final argument guarantees a default value.
+__typename: kind->match(
+    ["dog", "Canine"],
+    ["cat", "Feline"],
+    ["Exotic"]
+)
```

Comment:
I'm not opposed to going back to `[@, "Exotic"]`, but during the time when I wasn't sure about reusing `$` or adding `@`, I realized I could still provide a useful `->match` method that avoids both `$` and `@` by making `["Exotic"]` meaningful.

Technically speaking, all methods take a variable-length sequence of `JSON` expressions, and have the freedom to interpret whatever gets passed in as they see fit. The irregularity of mixing `[<key>, <value>]` pairs with `[<default>]` single-item arrays is no problem for a method that chooses to provide that kind of contract.

I think it's important for language simplicity/learnability that method arguments resemble JSON literals, except for the addition of variable references (`$` and `@` paths). That's why I'd rather not invent new `=>` syntax for method arguments, and I'd rather not invent a whole new `match` syntax when we can reuse the existing `->` method system. Syntax isn't cheap, both because there are only so many dedicated symbols you can use, and because learning a bunch of symbols is more difficult than learning a vocabulary of named methods. When you can achieve a whole bunch of use cases with one generic syntax (e.g. `->` methods with JSON-like arguments), that's powerful.

These animal examples may not show it off, but I fully intended for `->match` to allow matching  JSON values that are not just string literals. I think the `[<key>, <value>]` syntax is still the clearest way to express that with a JSON-like syntax.

Reply:
Ah, interesting. I wasn't thinking about it in terms of JSON literals at all, but rather specific-to-our-language features. I guess the thing we need to decide is where to encode the complexity—in the syntax of the language or in the levels of validation or tooling above it.

If we leave the language syntax as, generally "any number of literals can be passed to any method", then it's up to the validator, language server, etc. to know the specific meaning of each and every individual method. How many arguments are allowed in each, is any type allowed or just some, do different arguments have different restrictions, etc. For syntax highlighting, should certain values (in this case the left side of the match) be treated differently than others to better encode semantics?

I suppose "learnability" is subjective, but I've always found it much easier to learn to use languages which have explicit types than ones which, effectively, pass `Any` everywhere.

Reply:
> I guess the thing we need to decide is where to encode the complexity—in the syntax of the language or in the levels of validation or tooling above it.

I would recommend thinking of the JSONSelection parser and runtime as generic and relatively unopinionated, so we don't have to keep changing things at this level when we decide we want different rules. It's not bytecode, but you could compare it to using untyped JavaScript as a target language for TypeScript compilation.

I fully agree a type system will help with reasoning about the language, and it will need to participate in the static analysis of input shapes and output shapes (not just dynamic runtime enforcement). In my mind, that's a matter of expressing a type signature for each method, though the simplest adequate way of expressing and processing those types is still an open question. I don't think the validation logic will need to "know the specific meaning of each and every individual method" beyond understanding the type signature, as long as the type system we implement captures enough of what we want to validate.

Without a type system of some sort, of course you're right that higher-level validation code would have to embed knowledge about individual methods in order to reason about their behavior, but I'm hoping to avoid that.

Reply:
I second @dylan-apollo here, as having a comprehensive type system won't solve the issue of having to have special handling for what appears to be a regular function. Please let me know if you end up having this discussion offline as it has direct implications on the expansion code.

Reply:
Removed the single-element `[<default>]` case from `->match` for now: https://github.com/apollographql/router/pull/5762/commits/7353f00a3780eb7f0812d44ee604ab773f90decb

However, I do have a plan for the type system that could handle method signatures of this kind.

The escape hatch is we don't have to come up with a new syntax (be it TypeScript-like or Rust-like or Python-like) to express method type signatures, because we can write a Rust function to compute the output shape of any given `input->method(arg, ...)` invocation, based on the shapes of the `input` and arguments.

We still need to define a common representation of input/output selection `Shape`s, but once we have that Rust data structure, we could define the "type signature" of the `->echo(value)` method with a Rust function like
```rs
fn echo_method_shape(
    input_shape: &Shape,
    arg_shapes: &Vec<Shape>,
) -> Shape {
    if let Some(first) = arg_shapes.first() {
        first.clone()
    } else {
        Shape::Error("->echo expects one argument")
    }
}
```
This gets a bit more complicated with `match_method_shape`, but in principle there's nothing challenging about a variable-length argument list, because each individual `->match` invocation receives a fixed sequence of arguments with statically known shapes, and `match_method_shape` gets to perform arbitrary logic on that information.

Rust is a powerful (Turing-complete) language, but we are not exposing that power to the JSONSelection language. Instead, we'd be allowing ourselves to use an existing language (Rust) to define `->` method type signatures, conveniently the same language we use to implement the methods and to represent/process `Shape` information.

Reply:
Marking this as resolved, since we discussed this method live and landed on the current design (with `[@, value]` for default)